### PR TITLE
feat(api): add missing api routes

### LINKED ISSUE #45: Missing API routes
- `GET /modmail/v1/guilds/:guildId/snippets`
- `PUT /modmail/v1/guilds/:guildId/snippets` - creates a new snippet, returning it

- `GET /modmail/v1/guilds/:guildId/snippets/:snippetId`
- `PATCH /modmail/v1/guilds/:guildId/snippets/:snippetId` - update an existing snippet - also needs to insert a [SnippetUpdate](https://github.com/ChatSift/ModMail/blob/9f713cdb1ea1e15282649f55ee0a1faffc410cfe/prisma/schema.prisma#L20-L27), returns the updated data
- `DELETE /modmail/v1/guilds/:guildId/snippets/:snippetId`, returns the deleted snippet

### PR DESCRIPTION
**Pull request by felipebulcao**
Adds all the missing endpoints described in the issue below.
Closes #45

### CODE REVIEW

**Review by Codeize — APPROVED** [review]

**Review by OfficialSirH — APPROVED** [review]
a work of art

Let's
Get
This
Moving!!!!!

File: packages/api/src/routes/createSnippet.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import { jsonParser, Route, RouteMethod } from '@chatsift/rest-utils';
+import { REST } from '@discordjs/rest';
+import { badRequest } from '@hapi/boom';
+import { PrismaClient } from '@prisma/client';
+import type { BaseValidator, InferType } from '@sapphire/shapeshift';
+import { s } from '@sapphire/shapeshift';
+import type {
+	RESTPostAPIApplicationCommandsResult,
+	RESTPostAPIApplicationGuildCommandsJSONBody,
+} from 'discord-api-types/v9';
```
Comment (didinele) [inline packages/api/src/routes/createSnippet.ts]:
```suggestion
} from 'discord-api-types/v10';
```

File: packages/api/src/routes/createSnippet.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import { ApplicationCommandOptionType, Routes } from 'discord-api-types/v9';
```
Comment (didinele) [inline packages/api/src/routes/createSnippet.ts]:
```suggestion
import { ApplicationCommandOptionType, Routes } from 'discord-api-types/v10';
```

File: packages/api/src/index.ts [hunk 3]
```diff
@@ -16,6 +17,7 @@ import { logger } from './util/logger';
 
 const env = container.resolve(Env);
 container.register(PrismaClient, { useValue: new PrismaClient() });
+container.register(REST, { useValue: new REST({ version: '9' }).setToken(env.discordToken) });
```
Comment (didinele) [inline packages/api/src/index.ts]: leave it as is; we want v10

```suggestion
container.register(REST, { useValue: new REST().setToken(env.discordToken) });
```

File: packages/api/src/routes/deleteSnippet.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
+import type { TRequest } from '@chatsift/rest-utils';
+import { jsonParser, Route, RouteMethod } from '@chatsift/rest-utils';
+import { REST } from '@discordjs/rest';
+import { badRequest, notFound } from '@hapi/boom';
+import { PrismaClient } from '@prisma/client';
+import type { BaseValidator, InferType } from '@sapphire/shapeshift';
+import { s } from '@sapphire/shapeshift';
+import { Routes } from 'discord-api-types/v9';
```
Comment (didinele) [inline packages/api/src/routes/deleteSnippet.ts]:
```suggestion
import { Routes } from 'discord-api-types/v10';
```

File: packages/api/src/routes/createSnippet.ts (hunk 1 above)
Comment (didinele) [inline packages/api/src/routes/createSnippet.ts]:
```suggestion
} from 'discord-api-types/v10';
```

File: packages/api/src/routes/createSnippet.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+@singleton()
+export default class extends Route<Snippet, Body> {
+	public info = {
+		method: RouteMethod.put,
+		path: '/modmail/v1/guilds/:guildId/snippets/',
+	} as const;
+
+	public override readonly bodyValidationSchema: BaseValidator<Body> = schema;
+
+	public override middleware: Middleware[] = [jsonParser()];
```
Comment (didinele) [inline packages/api/src/routes/createSnippet.ts]: `jsonParser()` is automatically added by `Route`'s constructor if there's a `bodyValidationSchema`, this is redundant.

File: packages/api/src/routes/createSnippet.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+		const { guildId } = req.params as { guildId: string };
+
+		const snippetAlreadyExists = await this.prisma.snippet.findFirst({
+			where: {
+				name: req.body.name,
+				guildId,
+			},
+		});
+		if (snippetAlreadyExists) {
+			throw badRequest('A snippet with this name already exists in the guild');
```
Comment (didinele) [inline packages/api/src/routes/createSnippet.ts]: Prefer you take in `next: NextHandler` in `handle`

```suggestion
			return next(badRequest('A snippet with this name already exists in the guild'));
```

File: packages/api/src/routes/deleteSnippet.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	) {
+		super();
+	}
+
+	public async handle(req: TRequest<Body>, res: Response) {
+		const { guildId, snippetId } = req.params as { guildId: string; snippetId: string };
+
+		const snippetIdNum = Number.parseInt(snippetId, 10);
+		if (Number.isNaN(snippetIdNum)) {
+			throw badRequest('Invalid snippet ID');
```
Comment (didinele) [inline packages/api/src/routes/deleteSnippet.ts]: also this

File: packages/api/src/routes/deleteSnippet.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+		}
+
+		const snippet = await this.prisma.snippet.findFirst({
+			where: {
+				snippetId: snippetIdNum,
+			},
+		});
+		if (!snippet) {
+			throw notFound('Snippet not found');
```
Comment (didinele) [inline packages/api/src/routes/deleteSnippet.ts]: and this

File: packages/api/src/routes/deleteSnippet.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+@singleton()
+export default class extends Route<Snippet, Body> {
+	public info = {
+		method: RouteMethod.delete,
+		path: '/modmail/v1/guilds/:guildId/snippets/:snippetId',
+	} as const;
+
+	public override readonly bodyValidationSchema: BaseValidator<Body> = schema;
+
+	public override middleware: Middleware[] = [jsonParser()];
```
Comment (didinele) [inline packages/api/src/routes/deleteSnippet.ts]: oh and this

File: packages/api/src/routes/getSnippet.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+	public constructor(private readonly prisma: PrismaClient) {
+		super();
+	}
+
+	public async handle(req: Request, res: Response) {
+		const { guildId, snippetId } = req.params as { guildId: string; snippetId: string };
+
+		const snippetIdNum = Number.parseInt(snippetId, 10);
+		if (Number.isNaN(snippetIdNum)) {
+			throw badRequest('Invalid snippet ID');
```
Comment (didinele) [inline packages/api/src/routes/getSnippet.ts]: use `next` instead

File: packages/api/src/routes/getSnippet.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+		}
+
+		const snippet = await this.prisma.snippet.findFirst({
+			where: {
+				guildId,
+				snippetId: snippetIdNum,
+			},
+		});
+		if (!snippet) {
+			throw notFound('Snippet not found');
```
Comment (didinele) [inline packages/api/src/routes/getSnippet.ts]: use `next` instead

File: packages/api/src/routes/updateSettings.ts [hunk 12]
```diff
@@ -26,6 +26,8 @@ export default class extends Route<GuildSettings, Body> {
 
 	public override readonly bodyValidationSchema: BaseValidator<Body> = schema;
 
+	public override middleware: Middleware[] = [jsonParser()];
```
Comment (didinele) [inline packages/api/src/routes/updateSettings.ts]: no need

File: packages/api/src/routes/updateSnippet.ts [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+@singleton()
+export default class extends Route<Snippet, Body> {
+	public info = {
+		method: RouteMethod.patch,
+		path: '/modmail/v1/guilds/:guildId/snippets/:snippetId',
+	} as const;
+
+	public override readonly bodyValidationSchema: BaseValidator<Body> = schema;
+
+	public override middleware: Middleware[] = [jsonParser()];
```
Comment (didinele) [inline packages/api/src/routes/updateSnippet.ts]: no need

File: packages/api/src/routes/updateSnippet.ts [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+	) {
+		super();
+	}
+
+	public async handle(req: TRequest<Body>, res: Response) {
+		const { guildId, snippetId } = req.params as { guildId: string; snippetId: string };
+
+		const snippetIdNum = Number.parseInt(snippetId, 10);
+		if (Number.isNaN(snippetIdNum)) {
+			throw badRequest('Invalid snippet ID');
```
Comment (didinele) [inline packages/api/src/routes/updateSnippet.ts]: use next instead

File: packages/api/src/routes/updateSnippet.ts [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+		}
+
+		const snippet = await this.prisma.snippet.findFirst({
+			where: {
+				snippetId: snippetIdNum,
+			},
+		});
+		if (!snippet) {
+			throw notFound('Snippet not found');
```
Comment (didinele) [inline packages/api/src/routes/updateSnippet.ts]: use next instead

File: packages/api/src/util/env.ts [hunk 16]
```diff
@@ -5,6 +5,10 @@ import { singleton } from 'tsyringe';
 export class Env {
 	public readonly port: number = Number.parseInt(process.env.PORT ?? '8080', 10);
 
+	public readonly discordToken: string = process.env.DISCORD_TOKEN!;
```
Comment (didinele) [inline packages/api/src/util/env.ts:8]: Please add `DISCORD_TOKEN` and `DISCORD_CLIENT_ID` here: 
https://github.com/ChatSift/ModMail/blob/dc09d57ab8a6c2855fad68f1f489d6efeb45758d/packages/api/src/util/env.ts#L16

Name's a little unintuitive, but those are "required" env vars.

**Review by didinele — CHANGES_REQUESTED** [review]
Other than those reviews I think everything is fine. I'll roll out a `@chatsift/rest-utils` today with proper behavior re the validation schema.

**Comment by didinele** [comment]
bump `@chatsift/rest-utils` to `0.6.2`

File: packages/api/src/util/env.ts (hunk 16 above)
Comment (felipebulcao) [reply packages/api/src/util/env.ts:8]: Done in 8c0eed6

File: packages/api/src/routes/updateSnippet.ts (hunk 15 above)
Comment (felipebulcao) [reply packages/api/src/routes/updateSnippet.ts]: Done in 57469db

File: packages/api/src/routes/updateSnippet.ts (hunk 14 above)
Comment (felipebulcao) [reply packages/api/src/routes/updateSnippet.ts]: Done in 57469db

File: packages/api/src/routes/createSnippet.ts (hunk 6 above)
Comment (felipebulcao) [reply packages/api/src/routes/createSnippet.ts]: Done in 57469db

File: packages/api/src/routes/deleteSnippet.ts (hunk 8 above)
Comment (felipebulcao) [reply packages/api/src/routes/deleteSnippet.ts]: Done in 57469db

File: packages/api/src/routes/deleteSnippet.ts (hunk 7 above)
Comment (felipebulcao) [reply packages/api/src/routes/deleteSnippet.ts]: Done in 57469db

File: packages/api/src/routes/updateSnippet.ts (hunk 13 above)
Comment (felipebulcao) [reply packages/api/src/routes/updateSnippet.ts]: Fixed in 948d6ab

File: packages/api/src/routes/updateSettings.ts (hunk 12 above)
Comment (felipebulcao) [reply packages/api/src/routes/updateSettings.ts]: Fixed in 948d6ab

File: packages/api/src/routes/deleteSnippet.ts (hunk 9 above)
Comment (felipebulcao) [reply packages/api/src/routes/deleteSnippet.ts]: Fixed in 948d6ab

File: packages/api/src/routes/createSnippet.ts (hunk 5 above)
Comment (felipebulcao) [reply packages/api/src/routes/createSnippet.ts]: Fixed in 948d6ab

File: packages/api/src/routes/getSnippet.ts (hunk 11 above)
Comment (felipebulcao) [reply packages/api/src/routes/getSnippet.ts]: Done in 57469db

File: packages/api/src/routes/getSnippet.ts (hunk 10 above)
Comment (felipebulcao) [reply packages/api/src/routes/getSnippet.ts]: Done in 57469db

File: packages/api/src/routes/createSnippet.ts (hunk 1 above)
Comment (felipebulcao) [reply packages/api/src/routes/createSnippet.ts]: Done in ffe7ae6

File: packages/api/src/routes/deleteSnippet.ts (hunk 4 above)
Comment (felipebulcao) [reply packages/api/src/routes/deleteSnippet.ts]: Done in ffe7ae6

File: packages/api/src/index.ts (hunk 3 above)
Comment (felipebulcao) [reply packages/api/src/index.ts]: Done in ffe7ae6

File: packages/api/src/routes/createSnippet.ts (hunk 2 above)
Comment (felipebulcao) [reply packages/api/src/routes/createSnippet.ts]: Done in ffe7ae6

File: packages/api/src/routes/createSnippet.ts (hunk 1 above)
Comment (felipebulcao) [reply packages/api/src/routes/createSnippet.ts]: Done in ffe7ae6

**Review by didinele — APPROVED** [review]

**Pull request merged**